File: semester 4/MPP/JFTS/FestivalNetworking/src/main/java/ro/mpp/net/server/FestivalClientJsonWorker.java
package ro.mpp.net.server;

import com.google.gson.Gson;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp.authenticator.IAuthenticator;
import ro.mpp.domain.Show;
import ro.mpp.domain.ShowArtist;
import ro.mpp.domain.Ticket;
import ro.mpp.net.dto.ShowArtistDTO;
import ro.mpp.net.dto.TicketDTO;
import ro.mpp.net.dto.UserDTO;
import ro.mpp.net.dto._netutils.DTOUtils;
import ro.mpp.net.protocol.*;
import ro.mpp.observer.IFestivalObserver;
import ro.mpp.observer.IFestivalService;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public class FestivalClientJsonWorker implements Runnable, IFestivalObserver {

    private final IFestivalService festivalService;
    private final Socket connection;

    private BufferedReader input;
    private PrintWriter output;
    private final Gson gson = new Gson();
    private volatile boolean running;

    private static final Logger logger = LogManager.getLogger(FestivalClientJsonWorker.class);

    public FestivalClientJsonWorker(IFestivalService festivalService, Socket connection) {
        logger.debug("Creating FestivalClientJsonWorker");
        this.festivalService = festivalService;
        this.connection = connection;

        try {
            input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            output = new PrintWriter(connection.getOutputStream());
            running = true;
        } catch (Exception e) {
            logger.error("Error initializing Json Worker: {}", e.getMessage());
        }
    }

    @Override
    public void run() {
        while (running) {
            try {
                String requestLine = input.readLine();
                if (requestLine == null) {       // client disconnected cleanly
                    running = false;
                    break;
                }
                logger.debug("Request received: {}", requestLine);
                Request request = gson.fromJson(requestLine, Request.class);
                Response response = this.handleRequest(request);
                if (response != null) {
                    this.sendResponse(response);
                }
            } catch (IOException e) {
                logger.error("Error reading request", e);
                running = false;
            }
        }
        closeConnection();
    }

    private Response handleRequest(Request request) {
        logger.debug("Handling request of type: {}", request.getRequestType());

        switch (request.getRequestType()) {
            case LOGIN: {
                UserDTO userDTO = request.getUser();
                logger.debug("Login request from: {}", userDTO.getUsername());

                var user = festivalService.authenticate(userDTO.getUsername(),  userDTO.getPassword());
                if (user.isEmpty()) {
                    return Response.builder()
                            .responseType(ResponseType.ERROR)
                            .errorMessage("User not found!")
                            .build();
                }
                festivalService.login(userDTO.getUsername(), userDTO.getPassword(), this);
                return Response.builder()
                        .responseType(ResponseType.OK)
                        .user(DTOUtils.getDTO(user.get()))
                        .build();

            }

            case LOGOUT: {
                logger.debug("Logout request");
                festivalService.logout(request.getUser().getUsername());
                running = false;
                return Response.builder()
                        .responseType(ResponseType.OK)
                        .build();
            }

            case FIND_ALL: {
                List<ShowArtist> all = festivalService.findAll();
                ShowArtistDTO[] DTOs = all.stream()
                        .map(DTOUtils::getDTO)
                        .toArray(ShowArtistDTO[]::new);

                return Response.builder()
                        .responseType(ResponseType.FIND_ALL)
                        .showArtists(DTOs)
                        .build();
            }

            case FIND_BY_DATE: {
                LocalDate date = LocalDate.parse(request.getDate());
                List<ShowArtist> filtered = festivalService.findByDate(date);
                ShowArtistDTO[] DTOs = filtered.stream()
                        .map(DTOUtils::getDTO)
                        .toArray(ShowArtistDTO[]::new);

                return Response.builder()
                        .responseType(ResponseType.FIND_BY_DATE)
                        .showArtists(DTOs)
                        .build();
            }

            case BUY_TICKET: {
                TicketDTO ticketDTO = request.getTicket();
                Show show = DTOUtils.getFromDTO(ticketDTO.getShow());

                Optional<Ticket> result = festivalService.sellTicket(show, ticketDTO.getBuyerName(), ticketDTO.getSeats());

                if (result.isPresent()) {
                    return Response.builder()
                            .responseType(ResponseType.OK)
                            .ticket(DTOUtils.getDTO(result.get()))
                            .build();
                }
                return Response.builder()
                        .responseType(ResponseType.ERROR)
                        .errorMessage("Could not sell ticket")
                        .build();
            }

            case MODIFY_TICKET: {
                if (festivalService.modifyTicket(request.getTicketId(), request.getSeats())) {
                    return Response.builder()
                            .responseType(ResponseType.OK)
                            .build();
                }
                return Response.builder()
                        .responseType(ResponseType.ERROR)
                        .errorMessage("Ticket not found")
                        .build();
            }

            default:
                logger.warn("Unknown request type: {}", request.getRequestType());
                return Response.builder()
                        .responseType(ResponseType.ERROR)
                        .errorMessage("Unknown request type")
                        .build();
        }
    }

    private void sendResponse(Response response) {
        String json = gson.toJson(response);
        logger.debug("Sending response: {}", json);
        synchronized (output) {
            output.println(json);
            output.flush();
        }
    }

    private void closeConnection() {
        try {
            input.close();
            output.close();
            connection.close();
            logger.debug("Connection closed");
        } catch (IOException e) {
            logger.error("Error closing connection", e);
        }
    }

    @Override
    public void ticketSold(Ticket ticket) {
        Response response = Response.builder()
                .responseType(ResponseType.TICKET_SOLD)
                .ticket(DTOUtils.getDTO(ticket))
                .build();
        sendResponse(response);
    }

    @Override
    public void ticketModified(Ticket ticket) {
        Response response = Response.builder()
                .responseType(ResponseType.TICKET_MODIFIED)
                .ticket(DTOUtils.getDTO(ticket))
                .build();
        sendResponse(response);
    }
}

File: semester 4/MPP/C#FTS/FTS.Networking/Networking/Service/FestivalClientJsonWorker.cs
using C_FTS.Domain;
using C_FTS.Networking.DTO;
using C_FTS.Networking.Protocol;

using log4net;

using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace FTS.Networking.Networking.Service
{
    public class FestivalClientJsonWorker : IFestivalObserver
    {
        private readonly IFestivalSubject _festivalService;
        private readonly TcpClient _client;

        private readonly NetworkStream _stream;
        private readonly StreamReader _reader;
        private readonly StreamWriter _writer;
        private volatile bool _running;

        private static readonly ILog logger = LogManager.GetLogger(typeof(FestivalClientJsonWorker));

        public FestivalClientJsonWorker(IFestivalSubject festivalService, TcpClient client)
        {
            logger.Debug("Creating FestivalClientJsonWorker");
            this._festivalService = festivalService;
            this._client = client;

            try
            {
                this._stream = client.GetStream();
                this._reader = new StreamReader(_stream, Encoding.UTF8);
                this._writer = new StreamWriter(_stream, Encoding.UTF8) { AutoFlush = true };
                this._running = true;
            }
            catch (Exception ex)
            {
                logger.Error("Error initializing Json Worker", ex);
            }
        }

        public void Run()
        {
            while (_running)
            {
                try
                {
                    string requestLine = _reader.ReadLine();
                    if (requestLine == null)
                    {
                        _running = false;
                        break;
                    }

                    logger.DebugFormat("Request received: {0}", requestLine);
                    Request request = JsonSerializer.Deserialize<Request>(requestLine);
                    Response response = HandleRequest(request);

                    if (response != null)
                        SendResponse(response);
                }
                catch (IOException ex)
                {
                    logger.Error("Error reading request", ex);
                    _running = false;
                }
            }
            CloseConnection();
        }

        private Response HandleRequest(Request request)
        {
            logger.DebugFormat("Handling request of type: {0}", request.Type);

            switch (request.Type)
            {
                case RequestType.LOGIN:
                    {
                        UserDTO userDTO = request.User;
                        logger.DebugFormat("Login request from: {0}", userDTO.Username);

                        var user = _festivalService.Authenticate(userDTO.Username, userDTO.Password);
                        if (user == null)
                        {
                            return new Response (
                                type: ResponseType.ERROR,
                                message: "User not found!"
                            );
                        }
                        _festivalService.Login(userDTO.Username, userDTO.Password, this);
                        return new Response(
                            type: ResponseType.OK, 
                            user: userDTO
                        );
                    }

                case RequestType.LOGOUT:
                    {
                        logger.Debug("Logout request");
                        _festivalService.Logout(request.User.Username);
                        _running = false;
                        return new Response(type: ResponseType.OK);
                    }

                case RequestType.FIND_ALL:
                    {
                        var all = _festivalService.FindAll();
                        ShowArtistDTO[] dtos = all.Select(DTOUtil.GetDTO).ToArray();
                        return new Response (
                            type: ResponseType.FIND_ALL,
                            showArtist: dtos
                        );
                    }

                case RequestType.FIND_BY_DATE:
                    {
                        DateTime date = DateTime.Parse(request.Date);
                        var filtered = _festivalService.FindByDate(date);
                        ShowArtistDTO[] dtos = filtered.Select(DTOUtil.GetDTO).ToArray();
                        return new Response(
                            type: ResponseType.FIND_BY_DATE,
                            showArtist: dtos
                        );
                    }

                case RequestType.BUY_TICKET:
                    {
                        TicketDTO ticketDTO = request.Ticket;
                        Show show = DTOUtil.GetFromDTO(ticketDTO.Show);

                        Ticket? result = _festivalService.SellTicket(show, ticketDTO.BuyerName, ticketDTO.NumberOfSeats);
                        if (result != null)
                        {
                            return new Response (
                                type: ResponseType.OK,
                                ticket: DTOUtil.GetDTO(result)
                            );
                        }
                        return new Response (
                            type: ResponseType.ERROR,
                            message: "Could not sell ticket"
                        );
                    }

                case RequestType.MODIFY_TICKET:
                    {
                        if (_festivalService.ModifyTicket(request.TicketId, request.Seats))
                        {
                            return new Response(type: ResponseType.OK);
                        }
                        return new Response (
                            type: ResponseType.ERROR,
                            message: "Ticket not found"
                        );
                    }

                default:
                    logger.WarnFormat("Unknown request type: {0}", request.Type);
                    return new Response(
                        type: ResponseType.ERROR,
                        message: "Unknown request type"
                    );
            }
        }

        private void SendResponse(Response response)
        {
            string json = JsonSerializer.Serialize(response);
            logger.DebugFormat("Sending response: {0}", response.Type.ToString());
            lock (_writer)
            {
                _writer.WriteLine(json);
            }
        }

        private void CloseConnection()
        {
            try
            {
                _reader.Close();
                _writer.Close();
                _stream.Close();
                _client.Close();
                logger.Debug("Connection closed");
            }
            catch (IOException ex)
            {
                logger.Error("Error closing connection", ex);
            }
        }

        public void TicketSold(Ticket ticket)
        {
            SendResponse(new Response(
                type: ResponseType.TICKET_SOLD,
                ticket: DTOUtil.GetDTO(ticket)
            ));
        }

        public void TicketModified(Ticket ticket)
        {
            SendResponse(new Response (
                type: ResponseType.TICKET_MODIFIED,
                ticket: DTOUtil.GetDTO(ticket)
            ));
        }
    }
}